File: bench/kernel.cpp
// The same work in C++, built the way a host of this language is built, so the
// two numbers are of one machine and one compiler. It answers with the same
// checksum, which is what says it ran the same work.
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <vector>

struct Body {
    double x, y, dx, dy;
};

int main(int argc, char **argv) {
    const int many = 20000;
    const int rounds = 100;
    (void)argc;
    (void)argv;
    std::vector<Body> world;
    world.reserve(many);
    for (int i = 0; i < many; i++) {
        double f = (double)i;
        world.push_back({fmod(f, 1000.0), fmod(f * 7.0, 1000.0),
                         1.0 + fmod(f, 3.0),
                         1.0 + fmod(f, 5.0)});
    }
    for (int r = 0; r < rounds; r++) {
        for (size_t at = 0; at < world.size(); at++) {
            Body one = world[at];
            one.x += one.dx;
            one.y += one.dy;
            if (one.x < 0.0 || one.x > 1000.0) {
                one.dx = -one.dx;
            }
            if (one.y < 0.0 || one.y > 1000.0) {
                one.dy = -one.dy;
            }
            world[at] = one;
        }
    }
    double sum = 0.0;
    for (const Body &one : world) {
        sum += one.x + one.y;
    }
    printf("kernel %.0f\n", sum);
    return 0;
}

File: bench/kernel.lua
-- The same work in Luau. A table of four numbers a body, walked and moved,
-- with the same checksum at the end.
local many = 20000
local rounds = 100

local x, y, dx, dy = {}, {}, {}, {}
for i = 0, many - 1 do
    local f = i + 0.0
    x[i] = f % 1000.0
    y[i] = (f * 7.0) % 1000.0
    dx[i] = 1.0 + f % 3.0
    dy[i] = 1.0 + f % 5.0
end

for _ = 1, rounds do
    for at = 0, many - 1 do
        local ax = x[at] + dx[at]
        local ay = y[at] + dy[at]
        if ax < 0.0 or ax > 1000.0 then dx[at] = -dx[at] end
        if ay < 0.0 or ay > 1000.0 then dy[at] = -dy[at] end
        x[at] = ax
        y[at] = ay
    end
end

local sum = 0.0
for at = 0, many - 1 do
    sum = sum + x[at] + y[at]
end
print(string.format("kernel %.0f", sum))
